### ISSUE TITLE
Mypy `unsupported operand types` after inline checking for is None in typing.Optional[int] arg

### ISSUE THREAD

**Issue by StickySli** [python/mypy#12275]
**Bug Report**

I am trying to declare an argument to be optional with the `typing` library. Since this argument can be either `int | None`, if I try to do a math expression with it I first have to check if it's `None`, otherwise an error would raise.

**To Reproduce**

Snippet of my code with code partially implemented:

```python
def foo(filename: str, size: Optional[int] = None):
    # open filename, if it doesn't exist, create it if size is not None and size is multiple  of someConstant
    # if file exists, read and get trueFileSize
    ...
    if size is not None and trueFileSize != size * someConstant:
        # if trueFileSize is different than size times someConstant, rewrites file here
        ...

```

**Actual Behavior**
As I understand, if an `and` logical operator first check returns `False`, the subsequent checks are ignored, like in C. If I'm correct, why `mypy` is telling me this
```
... error: Unsupported operand types for * ("int" and "None")
```

**Workaround**

I get that rewriting the code in two if statements solves the problem, but why `mypy` can't deal with the previous snippet but doesn't raise an error with the following snippet?

```python
def foo(filename: str, size: Optional[int] = None):
    # open filename, if it doesn't exist, create it size is not None
    # if file exists, read and get trueFileSize
    ...
    if size is not None:
        if trueFileSize != size * someConstant: # mypy: doesn't raise error
            # if trueFileSize is different than size times someConstant, rewrite file
            ...
```
Similar to https://mypy.readthedocs.io/en/stable/kinds_of_types.html#strict-optional

**My Environment**
- Mypy version used: 0.931
- Mypy command-line flags: --python-version 3.10
- Operating system and version: Windows 11 Pro Version 10.0.22000 Build 22000

**Comment by JelleZijlstra** [comment]
I can't reproduce this: https://mypy-play.net/?mypy=latest&python=3.10&gist=7eb428bf3122985da9a0b0bc04020011. Please give a concrete code sample that shows the bug you're seeing.
